### PR TITLE
Fix casting in TLVReader::Get to be safe.

### PR DESCRIPTION
**Pull request by bzbarsky-apple**
Casts of out-of-range values to signed integer types are not defined
by the C or C++ standards, but this code was expecting them to act as
a cast to the corresponding-sized unsigned integer type, then the
"undoing" of a signed-to-unsigned cast.

We instead introduce a CastToSigned() function that explicitly and
safely undoes a signed-to-unsigned cast and use that in TLVReader::Get
to eliminate the implementation-defined behavior dependence.

 #### Problem
If you have a `uint64_t` with value `255`, casting it to `int8_t` is not guaranteed to produce `-1`, but the `TLVReader` code depends on this behavior.

 #### Summary of Changes
Replace the implementation-dependent casting with a safe function that produces the desired results.

### CODE REVIEW

**Review by saurabhst — APPROVED** [review]

**Review by rwalker-apple — APPROVED** [review]

**Comment by rwalker-apple** [comment]
@mspang @andy31415 ?

**Comment by mspang** [comment]
Do we need to do this if c++ 20 will standardize two's complement representation for signed integers which is already de facto standard?

[1] https://stackoverflow.com/questions/55468807/does-c20-well-define-left-shift-for-signed-integers-that-overflow

File: src/lib/support/SafeInt.h [hunk 1]
```diff
@@ -105,6 +105,38 @@ bool CanCastTo(U arg)
...
+ *
+ * If the argument is small enough to be representable as a positive signed
+ * integer, returns that integer.  Otherwise, returns a negative integer which
+ * would, if cast to the type of the argument, produce the given value.
+ *
+ * So for example, if a uint8_t with value 254 is passed in this function will
+ * return an int8_t with value -2.
+ */
+template <typename T>
+typename std::enable_if<std::is_unsigned<T>::value, typename std::make_signed<T>::type>::type CastToSigned(T arg)
```
Comment (andy31415) [inline src/lib/support/SafeInt.h:119]: Could we have some unit tests for this? They would serve as examples as well.

**Review by andy31415 — APPROVED** [review]
We should add some unit tests for this TLVReader ... if the functions were unsafe/not working, hopefully unit tests would have caught that. Or if working by chance, unit tests would have shown that changes did not affect functionality.

Approving as unit tests could be done as a followup, however I believe we should soon.

**Comment by andy31415** [comment]
Could you also rebase to make sure bloatcheck sees the impact on code size? Hopefully there is none.

**Comment by bzbarsky-apple** [comment]
> [1] http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2018/p0907r1.html

If this happens, then yes, the behavior here will be well-defined and we would not need this function.

File: src/lib/support/SafeInt.h (hunk 1 above)
Comment (bzbarsky-apple) [reply src/lib/support/SafeInt.h:119]: Yes, will add tests.

**Comment by bzbarsky-apple** [comment]
Fwiw, we do have `TLVReader` tests at `src/lib/core/tests/TestCHIPTLV.cpp`.  It's hard for me to tell offhand whether they exercise the interesting edge cases in this code...

**Comment by bzbarsky-apple** [comment]
As far as bloat check goes, https://github.com/project-chip/connectedhomeip/runs/1221097449?check_suite_focus=true says no changes for all the artifacts for this PR.  I'm not sure whether we should try to change the bloat check script to add a single "no bloat changes" comment or something....

**Comment by andy31415** [comment]
> Fwiw, we do have `TLVReader` tests at `src/lib/core/tests/TestCHIPTLV.cpp`. It's hard for me to tell offhand whether they exercise the interesting edge cases in this code...

Did not see them ... was searching for TestTLVReader which matches the pattern I saw for other unit tests. Maybe low priority task to rename things to be consistent. I feel better that we actually have tests.

**Comment by bzbarsky-apple** [comment]
Yeah, the naming is interesting.  `TestCHIPTLV` seems to test both the reader and the writer..
